MTCannon: Fix `AssertEVMReverts` to correctly construct data

joohhnnn opened the pull request:
First part of https://github.com/ethereum-optimism/optimism/issues/12198

1. Added `WithRegisters` for initializing state's `Registers`.
2. Added `ThreadProofEncoder` helper function to generate `ThreadProof`.
3. Updated `AssertEVMReverts` to include `statePCs` for multiple memory requirements, added `ProofData` parameter for passing `ThreadProof`, and introduced `expectedReason` parameter for more precise testing.
4. Revised `TestEVMFault` and `TestEVM_UnsupportedSyscall` test functions to be compatible with `AssertEVMReverts`.
5. used hardcoding for the `illegal instruction` that needs `SecondMemoryProof`

mbaxter commented on cannon/mipsevm/iface.go:
```diff
@@ -56,6 +56,9 @@ type FPVMState interface {
 	// EncodeWitness returns the witness for the current state and the state hash
 	EncodeWitness() (witness []byte, hash common.Hash)
 
+	// EncodeProof returns the proof for the current state
+	EncodeProof() []byte
```
I probably wouldn't modify the main interface for this - in particular because the tests here need an "initial"/incomplete proof without the memory proofs that are generated by running FPVM.step().   

I'd suggest adding a helper function in the test utils [here](https://github.com/ethereum-optimism/optimism/blob/cd1cc1039b58ce5146b49d1a018bc303cdaa56b2/cannon/mipsevm/tests/helpers.go#L53-L59).

joohhnnn replied:
solved and replaced `EncodeProof`  with `ThreadProofEncoder`

mbaxter commented on cannon/mipsevm/testutil/mips.go:
```diff
@@ -182,11 +182,11 @@ func ValidateEVM(t *testing.T, stepWitness *mipsevm.StepWitness, step uint64, go
 
 // AssertEVMReverts runs a single evm step from an FPVM prestate and asserts that the VM panics
 func AssertEVMReverts(t *testing.T, state mipsevm.FPVMState, contracts *ContractMetadata, tracer *tracing.Hooks) {
-	insnProof := state.GetMemory().MerkleProof(state.GetPC())
```
I'd suggest accepting an expected revert message here.  Then you can check it with:
```
retVal, _, err := env.Call(vm.AccountRef(sender), contracts.Addresses.MIPS, input, startingGas, common.U2560)
require.Contains(t, string(retVal), expectedRevertMsg)
```

mbaxter commented on cannon/mipsevm/singlethreaded/state.go:
```diff
@@ -185,6 +185,15 @@ func (s *State) EncodeWitness() ([]byte, common.Hash) {
+	proofData := make([]byte, 0)
+	insnProof := s.Memory.MerkleProof(s.Cpu.PC)
+	proofData = append(proofData, insnProof[:]...)
+	//Fake SecondMemoryProof
+	proofData = append(proofData, insnProof[:]...)
+	return proofData
```
We can't generate fake data in production code - if necessary, this should be a test utility.

joohhnnn replied:
Yeah, I realized that not all situations require a SecondMemoryProof, so I've made it optional. Also, I used hardcoding for the illegal instruction that needs SecondMemoryProof, but I might optimize this later.

mbaxter reviewed (changes requested):
Thanks for raising this issue - great find!

joohhnnn committed:
1. Added `WithRegisters` for initializing state's `Registers`.
2. Added `ThreadProofEncoder` helper function to generate `ThreadProof`.
3. Updated `AssertEVMReverts` to include `statePCs` for multiple memory requirements, added `ProofData` parameter for passing `ThreadProof`, and introduced `expectedReason` parameter for more precise testing.
4. Revised `TestEVMFault` and `TestEVM_UnsupportedSyscall` test functions to be compatible with `AssertEVMReverts`.

joohhnnn committed:
lint fix and comment fix

mbaxter commented on cannon/mipsevm/multithreaded/testutil/mutators.go:
```diff
@@ -93,3 +93,7 @@ func (m *StateMutatorMultiThreaded) SetPreimageOffset(val uint32) {
+
+func (m *StateMutatorMultiThreaded) SetRegisters(index uint32, val uint32) {
+	m.state.GetRegistersRef()[index] = val
+}
```
We don't actually need an explicit mutator for the registers, because they are already mutable via this `GetRegistersRef` that is globally available on the `FPVMState` interface.  

Made a suggestion below on how to modify the new proof test helper so we can avoid having to inject the `WithRegisters` helper which should allow us to drop this function.

mbaxter commented on cannon/mipsevm/tests/helpers.go:
```diff
@@ -50,31 +50,50 @@ func multiThreadElfVmFactory(t require.TestingT, elfFile string, po mipsevm.Prei
 	return fpvm
 }
 
+type ThreadProofEncoder func(po mipsevm.PreimageOracle, stdOut, stdErr io.Writer, log log.Logger, opts ...testutil.StateOption) []byte
```
I think we could simplify a bit and make this new util more generally useful if we just accept a state directly and generate the whole proof: 
```suggestion
type ProofGenerator func(t require.TestingT, state FPVMState, memoryProofAddresses ...uint32) []byte
```

For example, the multithreaded one could look something like:
```
func multiThreadedProofGenerator(t require.TestingT, state FPVMState, memoryProofAddresses ...uint32) []byte {
    mtState, ok := state.(*multithreaded.State)
    if !ok {
        require.Fail(t, "Failed to cast FPVMState to multithreaded State type")
    }

    proofData := mtState.EncodeThreadProof()
    insnProof := mtState.GetMemory().MerkleProof(mtState.GetPC())
    proofData = append(proofData, insnProof[:]...)

    for _, addr := range memoryProofAddresses {
        memProof := mtState.GetMemory().MerkleProof(addr)
        proofData = append(proofData, memProof)
    }

    return proofData
}
```

joohhnnn replied:
nice!

mbaxter commented on cannon/mipsevm/tests/evm_common_test.go:
```diff
@@ -499,24 +499,31 @@ func TestEVMFault(t *testing.T) {
+				// Future improvements might involve having the Go VM indicate attempted memory accesses
+				// and caching this info to generate proofs for the solidity VM.
+				if tt.errMsg == "invalid instruction" {
+					statePCs = append(statePCs, 0xa7ef00cc)
```
This looks like `EndAddr - 4`?  Where does this value come in?

joohhnnn replied:
Aha this is a tricky case, I got this from @Inphi   

>"The EVM reverts from here - https://github.com/ethereum-optimism/optimism/blob/1217d4ac74e4dfddc23391f41e50e51f7c64a772/packages/contracts-bedrock/src/cannon/libraries/MIPSInstructions.sol#L125
Basically, the instruction decoder is sorta  buggy. It admits invalid instructions and attempts to perform a memory read for the 0xff opcode.
So if you were to provide an additional  valid memory proof, it should eventually reafch the codepath that detects an invalid instruction.
Basically, all opcodes >= 0x20 are readed as memory access opcodes. And we provide a 0xff for that test
The way to go here would be to append a stubbed, but valid, memory proof for some random address"

mbaxter replied:
ah - thanks for the explanation!

mbaxter commented on cannon/mipsevm/testutil/mips.go:
```diff
@@ -181,21 +182,37 @@ func ValidateEVM(t *testing.T, stepWitness *mipsevm.StepWitness, step uint64, go
+	ret, _, err := env.Call(vm.AccountRef(sender), contracts.Addresses.MIPS, input, startingGas, common.U2560)
 	require.EqualValues(t, err, vm.ErrExecutionReverted)
+
+	if expectedReason != nil {
+		reason := ""
+		if len(ret) > 4 { // 4 bytes for the error selector
+			unpacked, decodeErr := abi.UnpackRevert(ret)
+			if decodeErr == nil {
+				reason = unpacked
+			}
+		}
```
nice 🎉

mbaxter commented on cannon/mipsevm/testutil/mips.go:
```diff
@@ -181,21 +182,37 @@ func ValidateEVM(t *testing.T, stepWitness *mipsevm.StepWitness, step uint64, go
+	ret, _, err := env.Call(vm.AccountRef(sender), contracts.Addresses.MIPS, input, startingGas, common.U2560)
 	require.EqualValues(t, err, vm.ErrExecutionReverted)
+
+	if expectedReason != nil {
```
Can we fail if `expectedReason == nil` to avoid false negatives in the future?

joohhnnn replied:
Make sense!

joohhnnn committed:
avoid false negatives

joohhnnn committed:
delete WithRegisters && create ProofGenerator

Inphi reviewed (changes requested):
This PR overlaps with https://github.com/ethereum-optimism/optimism/pull/12082. See please refer to my comments on the other PR as they apply here.

joohhnnn committed:
assert multi outputs

joohhnnn committed:
fix naming issue

mbaxter commented on cannon/mipsevm/tests/evm_multithreaded_test.go:
```diff
@@ -1179,10 +1179,14 @@ func TestEVM_UnsupportedSyscall(t *testing.T) {
 			state.Memory.SetMemory(state.GetPC(), syscallInsn)
 			state.GetRegistersRef()[2] = syscallNum
-
+			proofData := state.EncodeThreadProof()
+			insnProof := state.GetMemory().MerkleProof(state.GetPC())
+			proofData = append(proofData, insnProof[:]...)
```
This can be simplified with your new helper:
```suggestion
			proofData := multiThreadedProofGenerator(t, state)
```

mbaxter commented on cannon/mipsevm/tests/helpers.go:
```diff
@@ -50,31 +50,73 @@ func multiThreadElfVmFactory(t require.TestingT, elfFile string, po mipsevm.Prei
+type ProofGenerator func(t require.TestingT, state mipsevm.FPVMState, memoryProofAddresses ...uint32) []byte
+
+func singalThreadedProofGenerator(t require.TestingT, state mipsevm.FPVMState, memoryProofAddresses ...uint32) []byte {
+	stState, ok := state.(*singlethreaded.State)
+	if !ok {
+		require.Fail(t, "Failed to cast FPVMState to singlethreaded State type")
+	}
+
```
No need to cast the state here since we're not using any methods specific to the singlethreaded implementation. 
```suggestion
```

mbaxter commented on cannon/mipsevm/tests/evm_common_test.go:
```diff
@@ -499,10 +499,11 @@ func TestEVMFault(t *testing.T) {
 		name   string
 		nextPC uint32
 		insn   uint32
+		errMsg string
```
I would configure the memory proof addresses in the test case instead of hardcoding the special case below (`tt.errMsg == "invalid instruction"`):
```suggestion
		errMsg string
		memoryProofAddresses []uint32
```

mbaxter commented on cannon/mipsevm/tests/evm_common_test.go:
```diff
@@ -515,8 +516,16 @@ func TestEVMFault(t *testing.T) {
+				// TODO: Hardcode here to achieve the expected result.
+				// Future improvements might involve having the Go VM indicate attempted memory accesses
+				// and caching this info to generate proofs for the solidity VM.
+				memoryProofAddress := uint32(0)
```
Looks like we don't need any proof address for the other test cases.  If you configure the `memoryProofAddresses` in the `cases` array we can drop this for clarity.

mbaxter reviewed (commented):
Looks good!!  Just a few last comments to tidy things up a bit and looks like there's a merge conflict that needs to be resolved.

joohhnnn committed:
Update cannon/mipsevm/tests/evm_multithreaded_test.go

Nice catch!

Co-authored-by: mbaxter <meredith.a.baxter@gmail.com>

joohhnnn committed:
Small fix

joohhnnn committed:
Merge branch 'develop' into 12198-1

joohhnnn committed:
link check

Inphi commented on cannon/mipsevm/testutil/mips.go:
```diff
@@ -182,21 +183,30 @@ func ValidateEVM(t *testing.T, stepWitness *mipsevm.StepWitness, step uint64, go
 // AssertEVMReverts runs a single evm step from an FPVM prestate and asserts that the VM panics
-func AssertEVMReverts(t *testing.T, state mipsevm.FPVMState, contracts *ContractMetadata, tracer *tracing.Hooks) {
-	insnProof := state.GetMemory().MerkleProof(state.GetPC())
+func AssertEVMReverts(t *testing.T, state mipsevm.FPVMState, contracts *ContractMetadata, tracer *tracing.Hooks, ProofData []byte, expectedReason *string) {
```
can we make `expectedReason` a value given the `NotNil` check on L201?

joohhnnn replied:
Sure and the `expectedReason` `NotNil `check deleted now.

joohhnnn committed:
make expectedReason a value & delete nil check

Inphi reviewed (approved):
Excellent! Thanks
